File: tests/src/test/javascript/userauth/UserIdTokenAuthenticationFactoryTest.js
/**
 * Copyright (c) 2014-2018 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * User ID token user authentication factory unit tests.
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
describe("UserIdTokenAuthenticationFactory", function() {
    var MslEncoderFormat = require('msl-core/io/MslEncoderFormat.js');
    var EntityAuthenticationScheme = require('msl-core/entityauth/EntityAuthenticationScheme.js');
    var UserAuthenticationScheme = require('msl-core/userauth/UserAuthenticationScheme.js');
    var UserIdTokenAuthenticationFactory = require('msl-core/userauth/UserIdTokenAuthenticationFactory.js');
    var UserIdTokenAuthenticationData = require('msl-core/userauth/UserIdTokenAuthenticationData.js');
    var MslEncodingException = require('msl-core/MslEncodingException.js');
    var MslError = require('msl-core/MslError.js');
    var MslUserAuthException = require('msl-core/MslUserAuthException.js');

    var MslTestConstants = require('msl-tests/MslTestConstants.js');
    var MockMslContext = require('msl-tests/util/MockMslContext.js');
    var MockAuthenticationUtils = require('msl-tests/util/MockAuthenticationUtils.js');
    var MslTestUtils = require('msl-tests/util/MslTestUtils.js');
    var MockMslUser = require('msl-tests/tokens/MockMslUser.js');
    var MockTokenFactory = require('msl-tests/tokens/MockTokenFactory.js');
    
    /** MSL encoder format. */
    var ENCODER_FORMAT = MslEncoderFormat.JSON;
    
    /** Key master token. */
    var KEY_MASTER_TOKEN = "mastertoken";
    
    /** MSL context. */
    var ctx;
    /** MSL encoder factory. */
    var encoder;
    /** Authentication utilities. */
    var authutils;
    /** User authentication factory. */
    var factory;
    /** Token factory. */
    var tokenFactory;
    
    /** Master token. */
    var MASTER_TOKEN;
    /** User ID token. */
    var USER_ID_TOKEN;

    beforeEach(function() {
        if (!ctx) {
            runs(function() {
                MockMslContext.create(EntityAuthenticationScheme.PSK, false, {
                    result: function(x) { ctx = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); },
                });
            });
            waitsFor(function() { return ctx; }, "ctx", MslTestConstants.TIMEOUT_CTX);
            runs(function() {
                encoder = ctx.getMslEncoderFactory();
                authutils = new MockAuthenticationUtils();
                factory = new UserIdTokenAuthenticationFactory(authutils);
                ctx.addUserAuthenticationFactory(factory);
                tokenFactory = new MockTokenFactory();
                ctx.setTokenFactory(tokenFactory);
            
                MslTestUtils.getMasterToken(ctx, 1, 1, {
                    result: function(x) { MASTER_TOKEN = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); },
                });
            });
            waitsFor(function() { return MASTER_TOKEN; }, "master token", MslTestConstants.TIMEOUT);
            runs(function() {
                var user = new MockMslUser("user1");
                MslTestUtils.getUserIdToken(ctx, MASTER_TOKEN, 1, user, {
                    result: function(x) { USER_ID_TOKEN = x; },
                    error: function(e) { expect(function() { throw e; }).not.toThrow(); },
                });
            });
            waitsFor(function() { return USER_ID_TOKEN; }, "user ID token", MslTestConstants.TIMEOUT);
        }
    });
    
    afterEach(function() {
        authutils.reset();
        tokenFactory.reset();
    });
    
    it("create data", function() {
        var data, userAuthMo;
        runs(function() {
            data = new UserIdTokenAuthenticationData(MASTER_TOKEN, USER_ID_TOKEN);
            data.getAuthData(encoder, ENCODER_FORMAT, {
                result: function(x) { userAuthMo = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return userAuthMo; }, "userAuthMo", MslTestConstants.TIMEOUT);

        var authdata;
        runs(function() {
            factory.createData(ctx, null, userAuthMo, {
                result: function(x) { authdata = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); }
            });
        });
        waitsFor(function() { return authdata; }, "authdata", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(authdata).not.toBeNull();
            expect(authdata instanceof UserIdTokenAuthenticationData).toBeTruthy();
            expect(authdata).toEqual(data);
        });
    });
    
    it("encode exception", function() {
        var userAuthMo;
        runs(function() {
            var data = new UserIdTokenAuthenticationData(MASTER_TOKEN, USER_ID_TOKEN);
            data.getAuthData(encoder, ENCODER_FORMAT, {
                result: function(x) { userAuthMo = x; },
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return userAuthMo; }, "userAuthMo", MslTestConstants.TIMEOUT);
        
        var exception;
        runs(function() {
            userAuthMo.remove(KEY_MASTER_TOKEN);
            factory.createData(ctx, null, userAuthMo, {
                result: function(x) {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);
        
        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslEncodingException(MslError.MSL_PARSE_ERROR));
        });
    });
    
    it("authenticate", function() {
        var data = new UserIdTokenAuthenticationData(MASTER_TOKEN, USER_ID_TOKEN);
        
        var user;
        runs(function() {
            factory.authenticate(ctx, MASTER_TOKEN.identity, data, null, {
                result: function(x) { user = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return user; }, "user", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(user).not.toBeNull();
            expect(user).toEqual(USER_ID_TOKEN.user);
        });
    });
    
    it("authenticate user ID token", function() {
        var masterToken;
        runs(function() {
            MslTestUtils.getMasterToken(ctx, MASTER_TOKEN.sequenceNumber + 1, MASTER_TOKEN.serialNumber + 1, {
                result: function(x) { masterToken = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return masterToken; }, "master token", MslTestConstants.TIMEOUT);
        
        var userIdToken;
        runs(function() {
            MslTestUtils.getUserIdToken(ctx, masterToken, USER_ID_TOKEN.serialNumber + 1, USER_ID_TOKEN.user, {
                result: function(x) { userIdToken = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return userIdToken; }, "user ID token", MslTestConstants.TIMEOUT);
        
        var user;
        runs(function() {
            var data = new UserIdTokenAuthenticationData(MASTER_TOKEN, USER_ID_TOKEN);
            factory.authenticate(ctx, MASTER_TOKEN.identity, data, userIdToken, {
                result: function(x) { user = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return user; }, "user", MslTestConstants.TIMEOUT);
        
        runs(function() {
            expect(user).toEqual(USER_ID_TOKEN.user);
        });
    });
    
    it("authenticate mismatched user ID token", function() {
        var masterToken;
        runs(function() {
            MslTestUtils.getMasterToken(ctx, 1, 1, {
                result: function(x) { masterToken = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return masterToken; }, "master token", MslTestConstants.TIMEOUT);
        
        var userIdToken;
        runs(function() {
            var user = new MockMslUser("user2");
            MslTestUtils.getUserIdToken(ctx, masterToken, 1, user, {
                result: function(x) { userIdToken = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return userIdToken; }, "user ID token", MslTestConstants.TIMEOUT);
        
        var exception;
        runs(function() {
            var data = new UserIdTokenAuthenticationData(MASTER_TOKEN, USER_ID_TOKEN);
            factory.authenticate(ctx, MASTER_TOKEN.identity, data, userIdToken, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);
        
        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslUserAuthException(MslError.USERIDTOKEN_USERAUTH_DATA_MISMATCH));
        });
    });
    
    it("untrusted master token", function() {
        var untrustedMasterToken;
        runs(function() {
            MslTestUtils.getUntrustedMasterToken(ctx, {
                result: function(x) { untrustedMasterToken = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return untrustedMasterToken; }, "untrusted master token", MslTestConstants.TIMEOUT);

        var exception;
        runs(function() {
            var data = new UserIdTokenAuthenticationData(untrustedMasterToken, USER_ID_TOKEN);
            factory.authenticate(ctx, MASTER_TOKEN.identity, data, null, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslUserAuthException(MslError.USERAUTH_MASTERTOKEN_NOT_DECRYPTED));
        });
    });

    it("mismatched master token identity", function() {
        var mismatchedCtx;
        runs(function() {
            MockMslContext.create(EntityAuthenticationScheme.X509, false, {
                result: function(x) { mismatchedCtx = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return mismatchedCtx; }, "mismatched ctx", MslTestConstants.TIMEOUT);
        
        var mismatchedMasterToken;
        runs(function() {
            MslTestUtils.getMasterToken(mismatchedCtx, 1, 1, {
                result: function(x) { mismatchedMasterToken = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return mismatchedMasterToken; }, "mismatched master token", MslTestConstants.TIMEOUT);

        var exception;
        runs(function() {
            var data = new UserIdTokenAuthenticationData(mismatchedMasterToken, USER_ID_TOKEN);
            factory.authenticate(ctx, MASTER_TOKEN.identity, data, null, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslUserAuthException(MslError.USERAUTH_ENTITY_MISMATCH));
        });
    });
    
    it("untrusted user ID token", function() {
        var untrustedUserIdToken;
        runs(function() {
            MslTestUtils.getUntrustedUserIdToken(ctx, MASTER_TOKEN, USER_ID_TOKEN.serialNumber, USER_ID_TOKEN.user, {
                result: function(x) { untrustedUserIdToken = x; },
                error: function(e) { expect(function() { throw e; }).not.toThrow(); },
            });
        });
        waitsFor(function() { return untrustedUserIdToken; }, "untrusted user ID token", MslTestConstants.TIMEOUT);

        var exception;
        runs(function() {
            var data = new UserIdTokenAuthenticationData(MASTER_TOKEN, untrustedUserIdToken);
            factory.authenticate(ctx, MASTER_TOKEN.identity, data, null, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslUserAuthException(MslError.USERAUTH_USERIDTOKEN_NOT_DECRYPTED));
        });
    });
    
    it("user not permitted", function() {
        var exception;
        runs(function() {
            authutils.disallowScheme(MASTER_TOKEN.identity, USER_ID_TOKEN.user, UserAuthenticationScheme.USER_ID_TOKEN);

            var data = new UserIdTokenAuthenticationData(MASTER_TOKEN, USER_ID_TOKEN);
            factory.authenticate(ctx, MASTER_TOKEN.identity, data, null, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslUserAuthException(MslError.USERAUTH_ENTITYUSER_INCORRECT_DATA));
        });
    });
    
    it("token revoked", function() {
        var exception;
        runs(function() {
            tokenFactory.setRevokedUserIdToken(USER_ID_TOKEN);

            var data = new UserIdTokenAuthenticationData(MASTER_TOKEN, USER_ID_TOKEN);
            factory.authenticate(ctx, MASTER_TOKEN.identity, data, null, {
                result: function() {},
                error: function(e) { exception = e; },
            });
        });
        waitsFor(function() { return exception; }, "exception", MslTestConstants.TIMEOUT);

        runs(function() {
            var f = function() { throw exception; };
            expect(f).toThrow(new MslUserAuthException(MslError.USERIDTOKEN_REVOKED));
        });
    });
});

File: tests/src/test/cpp/userauth/UserIdTokenAuthenticationFactoryTest.cpp
/**
 * Copyright (c) 2016-2018 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include <io/MslEncoderFactory.h>
#include <io/MslEncoderFormat.h>
#include <keyx/KeyExchangeScheme.h>
#include <MslEncodingException.h>
#include <MslUserAuthException.h>
#include <tokens/MasterToken.h>
#include <tokens/UserIdToken.h>
#include <userauth/EmailPasswordAuthenticationFactory.h>
#include <userauth/UserAuthenticationFactory.h>
#include <userauth/UserIdTokenAuthenticationData.h>
#include <userauth/UserIdTokenAuthenticationFactory.h>
#include <memory>
#include <string>

#include "../util/MockAuthenticationUtils.h"
#include "../util/MockMslContext.h"
#include "../util/MslTestUtils.h"
#include "../tokens/MockMslUser.h"
#include "../tokens/MockTokenFactory.h"

using namespace std;
using namespace testing;
using namespace netflix::msl;
using namespace netflix::msl::entityauth;
using namespace netflix::msl::keyx;
using namespace netflix::msl::io;
using namespace netflix::msl::tokens;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace userauth {

namespace {

/** Key master token. */
const string KEY_MASTER_TOKEN = "mastertoken";

} // namespace anonymous

/**
 * User ID token user authentication factory unit tests.
 */
class UserIdTokenAuthenticationFactoryTest : public ::testing::Test
{
public:
    UserIdTokenAuthenticationFactoryTest()
    : ctx(make_shared<MockMslContext>(EntityAuthenticationScheme::PSK, false))
    , encoder(ctx->getMslEncoderFactory())
    , authutils(make_shared<MockAuthenticationUtils>())
    , factory(make_shared<UserIdTokenAuthenticationFactory>(authutils))
    , tokenFactory(make_shared<MockTokenFactory>())
    , ENCODER_FORMAT(MslEncoderFormat::JSON)
    , MASTER_TOKEN(MslTestUtils::getMasterToken(ctx, 1L, 1L))
    {
        ctx->addUserAuthenticationFactory(factory);
        ctx->setTokenFactory(tokenFactory);
        shared_ptr<MslUser> user = make_shared<MockMslUser>(1);
        USER_ID_TOKEN = MslTestUtils::getUserIdToken(ctx, MASTER_TOKEN, 1L, user);
    }
    ~UserIdTokenAuthenticationFactoryTest()
    {
        authutils->reset();
        tokenFactory->reset();
    }
protected:
    /** MSL context. */
    shared_ptr<MockMslContext> ctx;
    /** MSL encoder factory-> */
    shared_ptr<MslEncoderFactory> encoder;
    /** Authentication utilities. */
    shared_ptr<MockAuthenticationUtils> authutils;
    /** User authentication factory-> */
    shared_ptr<UserAuthenticationFactory> factory;
    /** Token factory. */
    shared_ptr<MockTokenFactory> tokenFactory;
    /** MSL encoder format. */
    const MslEncoderFormat ENCODER_FORMAT;
    /** Master token. */
    shared_ptr<MasterToken> MASTER_TOKEN;
    /** User ID token. */
    shared_ptr<UserIdToken> USER_ID_TOKEN;
};

TEST_F(UserIdTokenAuthenticationFactoryTest, createData)
{
    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(MASTER_TOKEN, USER_ID_TOKEN);
    shared_ptr<MslObject> userAuthMo = data->getAuthData(encoder, ENCODER_FORMAT);

    shared_ptr<UserAuthenticationData> authdata = factory->createData(ctx, shared_ptr<MasterToken>(), userAuthMo);
    EXPECT_TRUE(authdata);
    EXPECT_TRUE(instanceof<UserIdTokenAuthenticationData>(authdata.get()));
    EXPECT_EQ(*data, *authdata);
}

TEST_F(UserIdTokenAuthenticationFactoryTest, encodeException)
{
//    thrown.expect(MslEncodingException.class);
//    thrown.expectMslError(MslError::MSL_PARSE_ERROR);

    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(MASTER_TOKEN, USER_ID_TOKEN);
    shared_ptr<MslObject> userAuthMo = data->getAuthData(encoder, ENCODER_FORMAT);
    userAuthMo->remove(KEY_MASTER_TOKEN);
    try {
        factory->createData(ctx, shared_ptr<MasterToken>(), userAuthMo);
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslEncodingException& e) {
        EXPECT_EQ(MslError::MSL_PARSE_ERROR, e.getError());
    }
}

TEST_F(UserIdTokenAuthenticationFactoryTest, authenticate)
{
    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(MASTER_TOKEN, USER_ID_TOKEN);
    shared_ptr<MslUser> user = factory->authenticate(ctx, MASTER_TOKEN->getIdentity(), data, shared_ptr<UserIdToken>());
    EXPECT_TRUE(user);
    EXPECT_EQ(USER_ID_TOKEN->getUser(), user);
}

TEST_F(UserIdTokenAuthenticationFactoryTest, authenticateUserIdToken)
{
    shared_ptr<MasterToken> masterToken = MslTestUtils::getMasterToken(ctx, MASTER_TOKEN->getSequenceNumber() + 1, MASTER_TOKEN->getSerialNumber() + 1);
    shared_ptr<UserIdToken> userIdToken = MslTestUtils::getUserIdToken(ctx, masterToken, USER_ID_TOKEN->getSerialNumber() + 1, USER_ID_TOKEN->getUser());
    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(MASTER_TOKEN, USER_ID_TOKEN);
    shared_ptr<MslUser> u = factory->authenticate(ctx, MASTER_TOKEN->getIdentity(), data, userIdToken);
    EXPECT_EQ(USER_ID_TOKEN->getUser(), u);
}

TEST_F(UserIdTokenAuthenticationFactoryTest, authenticateMismatchedUserIdToken)
{
//    thrown.expect(MslUserAuthException.class);
//    thrown.expectMslError(MslError::USERIDTOKEN_USERAUTH_DATA_MISMATCH);

    shared_ptr<MasterToken> masterToken = MslTestUtils::getMasterToken(ctx, 1, 1);
    shared_ptr<MslUser> user = make_shared<MockMslUser>(2);
    shared_ptr<UserIdToken> userIdToken = MslTestUtils::getUserIdToken(ctx, masterToken, 1, user);
    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(MASTER_TOKEN, USER_ID_TOKEN);
    try {
        factory->authenticate(ctx, MASTER_TOKEN->getIdentity(), data, userIdToken);
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslUserAuthException& e) {
        EXPECT_EQ(MslError::USERIDTOKEN_USERAUTH_DATA_MISMATCH, e.getError());
    }
}

TEST_F(UserIdTokenAuthenticationFactoryTest, untrustedMasterToken)
{
//    thrown.expect(MslUserAuthException.class);
//    thrown.expectMslError(MslError::USERAUTH_MASTERTOKEN_NOT_DECRYPTED);

    shared_ptr<MasterToken> untrustedMasterToken = MslTestUtils::getUntrustedMasterToken(ctx);
    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(untrustedMasterToken, USER_ID_TOKEN);
    try {
        factory->authenticate(ctx, MASTER_TOKEN->getIdentity(), data, shared_ptr<UserIdToken>());
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslUserAuthException& e) {
        EXPECT_EQ(MslError::USERAUTH_MASTERTOKEN_NOT_DECRYPTED, e.getError());
    }
}

TEST_F(UserIdTokenAuthenticationFactoryTest, mismatchedMasterTokenIdentity)
{
//    thrown.expect(MslUserAuthException.class);
//    thrown.expectMslError(MslError::USERAUTH_ENTITY_MISMATCH);

    shared_ptr<MslContext> mismatchedCtx = make_shared<MockMslContext>(EntityAuthenticationScheme::PSK_PROFILE, false);
    shared_ptr<MasterToken> mismatchedMasterToken = MslTestUtils::getMasterToken(mismatchedCtx, 1, 1);
    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(mismatchedMasterToken, USER_ID_TOKEN);
    try {
        factory->authenticate(ctx, MASTER_TOKEN->getIdentity(), data, shared_ptr<UserIdToken>());
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslUserAuthException& e) {
        EXPECT_EQ(MslError::USERAUTH_ENTITY_MISMATCH, e.getError());
    }
}

TEST_F(UserIdTokenAuthenticationFactoryTest, untrustedUserIdToken)
{
//    thrown.expect(MslUserAuthException.class);
//    thrown.expectMslError(MslError::USERAUTH_USERIDTOKEN_NOT_DECRYPTED);

    shared_ptr<UserIdToken> untrustedUserIdToken = MslTestUtils::getUntrustedUserIdToken(ctx, MASTER_TOKEN, USER_ID_TOKEN->getSerialNumber(), USER_ID_TOKEN->getUser());
    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(MASTER_TOKEN, untrustedUserIdToken);
    try {
        factory->authenticate(ctx, MASTER_TOKEN->getIdentity(), data, shared_ptr<UserIdToken>());
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslUserAuthException& e) {
        EXPECT_EQ(MslError::USERAUTH_USERIDTOKEN_NOT_DECRYPTED, e.getError());
    }
}

TEST_F(UserIdTokenAuthenticationFactoryTest, userNotPermitted)
{
//    thrown.expect(MslUserAuthException.class);
//    thrown.expectMslError(MslError::USERAUTH_ENTITYUSER_INCORRECT_DATA);

    authutils->disallowScheme(MASTER_TOKEN->getIdentity(), USER_ID_TOKEN->getUser(), UserAuthenticationScheme::USER_ID_TOKEN);

    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(MASTER_TOKEN, USER_ID_TOKEN);
    try {
        factory->authenticate(ctx, MASTER_TOKEN->getIdentity(), data, shared_ptr<UserIdToken>());
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslUserAuthException& e) {
        EXPECT_EQ(MslError::USERAUTH_ENTITYUSER_INCORRECT_DATA, e.getError());
    }
}

TEST_F(UserIdTokenAuthenticationFactoryTest, tokenRevoked)
{
//    thrown.expect(MslUserAuthException.class);
//    thrown.expectMslError(MslError.USERIDTOKEN_REVOKED);

    tokenFactory->setRevokedUserIdToken(USER_ID_TOKEN);

    shared_ptr<UserIdTokenAuthenticationData> data = make_shared<UserIdTokenAuthenticationData>(MASTER_TOKEN, USER_ID_TOKEN);
    try {
        factory->authenticate(ctx, MASTER_TOKEN->getIdentity(), data, USER_ID_TOKEN);
        ADD_FAILURE() << "Should have thrown.";
    } catch (const MslUserAuthException& e) {
        EXPECT_EQ(MslError::USERIDTOKEN_REVOKED, e.getError());
    }
}

}}} // namespace netflix::msl::userauth

